contracts/yellow/VaultImplBase.sol

```diff
@@ -0,0 +1,128 @@
+//SPDX-License-Identifier: MIT
+pragma solidity 0.8.4;
+
+import '@openzeppelin/contracts/interfaces/draft-IERC1822.sol';
+import '@openzeppelin/contracts/proxy/ERC1967/ERC1967Upgrade.sol';
+
+/**
+ * @dev Base logic for the Implementation, containing the `upgradeTo` and `upgradeToAndCall` methods.
+ */
+abstract contract VaultImplBase is IERC1822Proxiable, ERC1967Upgrade {
+  // VaultImplBase specific
+
+  constructor() {
+    ERC1967Upgrade._changeAdmin(msg.sender);
+  }
+
+  address private immutable __self = address(this);
+
+  /**
+    * @dev Check that the execution is being performed through a delegatecall call and that the execution context is
+    * a proxy contract with an implementation (as defined in ERC1967) pointing to self. This should only be the case
+    * for UUPS and transparent proxies that are using the current contract as their implementation.
+    */
+  modifier onlyProxy() {
+      require(address(this) != __self, "must be called through delegatecall");
+      require(_getImplementation() == __self, "must be called through active proxy");
+      _;
+  }
+
+  /**
+    * @dev Check that the execution is not being performed through a delegate call. This allows a function to be
+    * callable on the implementing contract but not through proxies.
+    */
+  modifier notDelegated() {
+      require(address(this) == __self, "must not be called through delegatecall");
+      _;
+  }
+
+  modifier onlyAdmin() {
+    require(msg.sender == ERC1967Upgrade._getAdmin(), 'caller not admin');
+    _;
+  }
+
+  // Implementation context storage
+
+  event NewerImplementationSet(address indexed newerImplementation);
+
+  /** @dev Double underscore enables using the same variable name with a single underscore in a derived contract */
+  address private __newerImplementation;
+
+  function getNewerImplementation() external view notDelegated returns (address) {
+    return __newerImplementation;
+  }
+
+  function setNewerImplementation(address newerImplementation) external notDelegated onlyAdmin {
+    require(__newerImplementation == address(0), 'newerImplementation is already set');
+    // prevent unnecessary event emitions & infinite newerImplementaion chain
+    require(
+      newerImplementation != address(0) && newerImplementation != __self,
+      'invalid newerImplementation supplied'
+    );
+    
+    __newerImplementation = newerImplementation;
+    
+    emit NewerImplementationSet(newerImplementation);
+  }
+
+  function getAdmin() external view returns (address) {
+    return ERC1967Upgrade._getAdmin();
+  }
+
+  function changeAdmin(address newAdmin) external onlyAdmin {
+    ERC1967Upgrade._changeAdmin(newAdmin);
+  }
```

Comment:
Can you move this above `// Implementation context storage`

In a specific `// Shared context` or something like this, to emphasis that this logic is shared for both Impl and Proxy contexts

Reply:
Will address this in dev docs commit

Reply:
7b2e745